lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp

```diff
@@ -1523,6 +1662,64 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
   }
 }
 
+// TODO: Consolidate with CSE and move this to InstOptUtils.cpp
+/// Given a newly created value 'Val', insert destroy_value at its frontiers
+/// when not already found
+static void fixupLifetimeOfValue(ValueLifetimeAnalysis &Vla, SILValue Val) {
+  ValueLifetimeAnalysis::Frontier ValFrontier;
+  Vla.computeFrontier(ValFrontier, ValueLifetimeAnalysis::DontModifyCFG);
+  // Go over all the frontiers and insert a destroy of the copy if a
+  // lifetime ending instruction was not found at the frontier
+  for (auto *End : ValFrontier) {
+    auto EndIt = End->getIterator();
+    SILInstruction *LastUser =
+        (EndIt == End->getParent()->begin()) ? nullptr : &*std::prev(EndIt);
+    // If the LastUser does not end the lifetime of Val, we have to insert a
+    // destroy_value of Val
+    bool HasLifetimeEnd = false;
+    if (LastUser) {
+      // If the LastUser is a begin_borrow, it is not lifetime ending, change
+      // the InsertPt to next instruction after end_borrow
+      if (isa<BeginBorrowInst>(LastUser)) {
+        auto EndBorrowRange = cast<BeginBorrowInst>(LastUser)->getEndBorrows();
+        // Since we never generate multiple end_borrows in RLE, we only handle
+        // the case where we have a single end_borrow here
+        assert(std::distance(EndBorrowRange.begin(), EndBorrowRange.end()) ==
+               1);
+        LastUser = *EndBorrowRange.begin();
+        End = &*std::next(LastUser->getIterator());
+      } else {
+        // Check if the LastUser is ending the lifetime of Val, if not we need
+        // to insert a destroy_value
+        auto LastUserOps = LastUser->getAllOperands();
+        HasLifetimeEnd = std::any_of(
+            LastUserOps.begin(), LastUserOps.end(), [&Val](Operand &Op) {
+              if (Op.get() == Val && Op.isLifetimeEnding())
+                return true;
+              return false;
+            });
+      }
+    } else {
+      // If the LastUser was a SILPhiArgument, check if any of the incoming
```

Comment:
If a frontier instruction is the first in a block, it does not necessarily mean that the last user is a terminal instruction in the predecessor block. E.g.
```
  %1 = def
loop_header:
    ... = use(%1)
   cond_br loop_header, loop_exit
loop_exit:
   // This is the frontier
```

Reply:
I missed this comment. 

Thanks for the example. But can such a use ever be consuming ?
In this non-consuming case, HasLifetimeEnd will be false (because we did not find it as an incoming value to the phi arg) and we will end up with a destroy_value correctly in the loop_exit.

Reply:
You are right, the code will work. But at least the comment is confusing, because the block might not have any phi arguments at all.

Reply:
Ok, I fixed the comment.